pkg/registry/admissionregistration/mutatingadmissionpolicybinding/authz.go

```diff
@@ -0,0 +1,117 @@
+	"k8s.io/kubernetes/pkg/apis/admissionregistration"
+	rbacregistry "k8s.io/kubernetes/pkg/registry/rbac"
+)
+
+func (v *mutatingAdmissionPolicyBindingStrategy) authorizeCreate(ctx context.Context, obj runtime.Object) error {
+	binding := obj.(*admissionregistration.MutatingAdmissionPolicyBinding)
+	if binding.Spec.ParamRef == nil {
+		// no paramRef in new object
+		return nil
+	}
+
+	return v.authorize(ctx, binding)
+}
+
+func (v *mutatingAdmissionPolicyBindingStrategy) authorizeUpdate(ctx context.Context, obj, old runtime.Object) error {
+	binding := obj.(*admissionregistration.MutatingAdmissionPolicyBinding)
+	if binding.Spec.ParamRef == nil {
+		// no paramRef in new object
+		return nil
+	}
+
+	oldBinding := old.(*admissionregistration.MutatingAdmissionPolicyBinding)
+	if oldBinding.Spec.ParamRef != nil && *oldBinding.Spec.ParamRef == *binding.Spec.ParamRef && oldBinding.Spec.PolicyName == binding.Spec.PolicyName {
+		// identical paramRef and policy to old object
+		return nil
+	}
+
+	return v.authorize(ctx, binding)
+}
+
+func (v *mutatingAdmissionPolicyBindingStrategy) authorize(ctx context.Context, binding *admissionregistration.MutatingAdmissionPolicyBinding) error {
+	if v.authorizer == nil || v.resourceResolver == nil || binding.Spec.ParamRef == nil {
+		return nil
+	}
+
+	// for superuser, skip all checks
+	if rbacregistry.EscalationAllowed(ctx) {
+		return nil
+	}
+
+	user, ok := genericapirequest.UserFrom(ctx)
+	if !ok {
+		return fmt.Errorf("cannot identify user to authorize read access to paramRef object")
+	}
+
+	// default to requiring permissions on all group/version/resources
+	resource, apiGroup, apiVersion := "*", "*", "*"
+
+	if policy, err := v.policyGetter.GetMutatingAdmissionPolicy(ctx, binding.Spec.PolicyName); err == nil && policy.Spec.ParamKind != nil {
+		paramKind := policy.Spec.ParamKind
+		if gv, err := schema.ParseGroupVersion(paramKind.APIVersion); err == nil {
+			// we only need to authorize the parsed group/version
+			apiGroup = gv.Group
+			apiVersion = gv.Version
+			if gvr, err := v.resourceResolver.Resolve(gv.WithKind(paramKind.Kind)); err == nil {
+				// we only need to authorize the resolved resource
+				resource = gvr.Resource
+			}
+		}
+	}
+
+	var attrs authorizer.AttributesRecord
+
+	paramRef := binding.Spec.ParamRef
+	verb := "get"
+
+	if len(paramRef.Name) == 0 {
+		verb = "list"
+	}
+
+	attrs = authorizer.AttributesRecord{
+		User:            user,
+		Verb:            verb,
+		ResourceRequest: true,
+		Name:            paramRef.Name,
+		Namespace:       paramRef.Namespace, // if empty, no namespace indicates get across all namespaces
+		APIGroup:        apiGroup,
+		APIVersion:      apiVersion,
+		Resource:        resource,
+	}
+
+	d, _, err := v.authorizer.Authorize(ctx, attrs)
+	if err != nil {
+		return err
+	}
+	if d != authorizer.DecisionAllow {
+		return fmt.Errorf(`user %v does not have "%v" permission on the object referenced by paramRef`, verb, user)
```

Comment:
with all the code paths above with missing policy, failed REST mapping, it's not super clear what was it that made this authorization fail. Should we be a little bit more verbose which permission we are actually checking? I.e. group, version, resource.

Reply:
I agree. It would be hard to know what to do as a user given only this error message.

I've expanded this to return different errors for these cases. I added tests to cover each.